docs/architecture/adr-058-auto-generated-cli.md

```diff
@@ -0,0 +1,83 @@
+# ADR 058: Auto-Generated CLI
+
+## Changelog
+
+* 2022-05-04: Initial Draft
+
+## Status
+
+ACCEPTED Partially Implemented
+
+## Abstract
+
+In order to make it easier for developers to write Cosmos SDK modules, we provide infrastructure which automatically
+generates CLI commands based on protobuf definitions.
+
+## Context
+
+Current Cosmos SDK modules generally implement a CLI command for every transaction and every query supported by the
+module. These are handwritten for each command and essentially amount to providing some CLI flags or positional
+arguments for specific fields in protobuf messages.
+
+In order to make sure CLI commands are correctly implemented as well as to make sure that the application works
+in end-to-end scenarios, we do integration tests using CLI commands. While these tests are valuable on some-level,
+they can be hard to write and maintain, and run slowly. [Some teams have contemplated](https://github.com/regen-network/regen-ledger/issues/1041)
+moving away from CLI-style integration tests (which are really end-to-end tests) towards narrower integration tests
+which exercise `MsgClient` and `QueryClient` directly. This might involve replacing the current end-to-end CLI
+tests with unit tests as there still needs to be some way to test these CLI commands for full quality assurance.
+
+## Decision
+
+To make module development simpler, we provide infrastructure - in the new [`client/v2`](https://github.com/cosmos/cosmos-sdk/tree/main/client/v2)
+go module - for automatically generating CLI commands based on protobuf definitions to either replace or complement
+handwritten CLI commands. This will mean that when developing a module, it will be possibly to skip both writing and
+testing CLI commands as that can all be taken care of by the framework.
+
+The basic design for automatically generating CLI commands is to:
+* create one CLI command for each `rpc` method in a protobuf `Query` or `Msg` service
+* create a CLI flag for each field in the `rpc` request type
```

Comment:
Since in most, if not all, cases fields will be required in order to correctly construct the relevant message and/or query, these really shouldn't be flags, but args.

That being said, a dev might want some as flags and not as args (i.e. they're optional and default values are totally OK). So can we add an annotation to the field that describes whether it'll be a flag or an arg?

Reply:
Perhaps pagination is an exception, in that it ought to be flags?

Reply:
Perhaps pagination is an exception, in that it ought to be flags?

Reply:
There are exceptions in a lot of Msgs. I propose to provide a sane default (i.e. flags IMO) here, and to allow customization separately.

Reply:
Is that not just defining a JSON configuration standard?
(And if so, lets just use standard YML standards for CLI's. This is how ~all the rust CLI's are built AFAIU)

Reply:
It's already there: Proto JSON